packages/web-pkg/src/components/AppTemplates/AppWrapper.vue

```diff
@@ -0,0 +1,339 @@
+    })
+
+    const hasProp = (name: string) => {
+      return Boolean(Object.keys(props.wrappedComponent.props).includes(name))
+    }
+
+    const isDirty = computed(() => {
+      return unref(currentContent) !== unref(serverContent)
+    })
+
+    const {
+      currentFileContext,
+      getFileContents,
+      getFileInfo,
+      getUrlForResource,
+      closeApp,
+      putFileContents,
+      revokeUrl,
+      replaceInvalidFileRoute
+    } = useAppDefaults({
+      applicationId: props.applicationId
+    })
+
+    const loadFileTask = useTask(function* () {
+      try {
+        // FIXME: make all updates atomic/happen at once so the ui can never mix data from different resources
+
+        const tmpResource = yield getFileInfo(currentFileContext, {
+          davProperties: [DavProperty.FileId, DavProperty.Permissions, DavProperty.Name]
+        })
+
+        replaceInvalidFileRoute(currentFileContext, tmpResource)
+
+        isReadOnly.value = ![DavPermission.Updateable, DavPermission.FileUpdateable].some(
+          (p) => (tmpResource.permissions || '').indexOf(p) > -1
+        )
+
+        if (unref(hasProp('currentContent'))) {
+          const fileContentsResponse = yield getFileContents(currentFileContext)
+          serverContent.value = currentContent.value = fileContentsResponse.body
+          currentETag.value = fileContentsResponse.headers['OC-ETag']
+        }
+
+        if (unref(hasProp('url'))) {
+          const currentSpace = unref(unref(currentFileContext).space)
+          const tmpUrl = yield getUrlForResource(
+            currentSpace,
+            tmpResource,
+            props.urlForResourceOptions
+          )
+          url.value = tmpUrl
+        }
+
+        resource.value = tmpResource
+      } catch (e) {
+        console.error(e)
+        loadingError.value = true
+      } finally {
+        loading.value = false
+      }
+    }).restartable()
+
+    watch(
+      currentFileContext,
+      () => {
+        loadFileTask.perform()
+      },
+      { immediate: true }
+    )
+
+    onBeforeUnmount(() => {
+      revokeUrl(url.value)
+    })
+
+    const errorPopup = (error) => {
+      store.dispatch('showErrorMessage', {
+        title: $gettext('An error occurred'),
+        desc: error,
+        error
+      })
+    }
+
+    const autosavePopup = () => {
+      store.dispatch('showMessage', {
+        title: $gettext('File autosaved')
+      })
+    }
+
+    const saveFileTask = useTask(function* () {
+      const newContent = unref(currentContent)
+      try {
+        const putFileContentsResponse = yield putFileContents(currentFileContext, {
+          content: newContent,
+          previousEntityTag: unref(currentETag)
+        })
+        serverContent.value = newContent
+        currentETag.value = putFileContentsResponse.etag
+      } catch (e) {
+        // FIXME: handle 403 error when trying to save a readonly file
```

Comment:
@JanAckermann for you <3

Reply:
will continue with that on Monday, if not merged

Reply:
Marking as resolved, thx @JanAckermann